src/zenml/utils/tag_utils.py

```diff
@@ -0,0 +1,636 @@
+            )
+
+        step_output_names = list(step_context._outputs.keys())
+
+        if artifact_name is not None:
+            # If a name provided, ensure it is in the outputs
+            if artifact_name not in step_output_names:
+                raise ValueError(
+                    f"The provided artifact name`{artifact_name}` does not "
+                    f"exist in the step outputs: {step_output_names}."
+                )
+        else:
+            # If no name provided, ensure there is only one output
+            if len(step_output_names) > 1:
+                raise ValueError(
+                    "There is more than one output. If you would like to use "
+                    "the `infer_artifact` option, you need to define an "
+                    "`artifact_name`."
+                )
+
+            if len(step_output_names) == 0:
+                raise ValueError("The step does not have any outputs.")
+
+            artifact_name = step_output_names[0]
+
+        step_context.add_output_tags(
+            tags=[t.name if isinstance(t, Tag) else t for t in tags],
+            output_name=artifact_name,
+        )
+
+    # If every additional value is None, that means we are calling it bare bones
+    # and this call needs to happen during a step execution. We will use the
+    # step context to fetch the run and attach the tags accordingly.
+    elif all(
+        v is None
+        for v in [
+            run,
+            artifact_version_id,
+            artifact_name,
+            artifact_version,
+            pipeline,
+            run_template,
+        ]
+    ):
+        try:
+            from zenml.steps.step_context import get_step_context
+
+            step_context = get_step_context()
+        except RuntimeError:
+            raise ValueError(
+                f"""
+                You are calling 'add_tags()' outside of a step execution. 
+                If you would like to add tags to a ZenML entity outside 
+                of the step execution, please provide the required 
+                identifiers.\n{add_tags_warning}
+                """
+            )
+
+        # Tag the pipeline run, not the step
+        resource_id = step_context.pipeline_run.id
+        resource_type = TaggableResourceTypes.PIPELINE_RUN
+
+    else:
+        raise ValueError(
+            f"""
+            Unsupported way to call the `add_tags`. Possible combinations "
+            include: \n{add_tags_warning}
+            """
+        )
+
+    # Create tag resources and add tags
+    for tag in tags:
+        try:
+            if isinstance(tag, Tag):
+                tag_model = client.get_tag(tag.name)
+
+                if tag.exclusive and resource_type not in [
+                    TaggableResourceTypes.PIPELINE_RUN,
+                    TaggableResourceTypes.ARTIFACT_VERSION,
+                    TaggableResourceTypes.RUN_TEMPLATE,
+                ]:
+                    raise ValueError(
```

Comment:
I don't get this error: Why can't I create a `Tag` and add it to a model let's say, but also make that tag exclusive? Sure it won't apply to my current entity, but it's still valid no?

Reply:
IMO, if we do not handle the condition here, it would create a false sense of functionality for the user. However, you are right on one thing. You should be able to attach an exclusive tag to an entity regardless of its type. So, I adjusted the code to throw a warning in case people want to use an exclusive tag with a type that does not support the functionality.